support building parts of Boost with Emscripten

TimSimpson opened the pull request:
This changes the Boost conanfile to not fail if the `arch` profile setting is `asm.js` and to explicitly use the `emscripten` Boost Build toolchain if the `os` profile setting is `Emscripten`.

Using this, I've been able to make Boost build with Emscripten- if I disable several components and avoid using certain libraries. To make this pass the package test I had to hack it up to make it avoid
using the regex libraries.

Here's the profile I'm using:

```
[settings]
os=Emscripten
arch=asm.js
os_build=Linux
arch_build=x86_64
compiler=clang
compiler.version=6.0
compiler.libcxx=libc++
compiler.cppstd=17

[options]
boost:without_graph=True
boost:without_locale=True
boost:without_log=True
boost:without_python=True
boost:without_regex=True
boost:without_stacktrace=True
boost:without_test=True
boost:zlib=False
boost:bzip2=False
boost:lzma=False
boost:zstd=False

[build_requires]
emsdk_installer/1.38.29@bincrafters/stable
```

If I don't disable the libraries in the profile, I get error messages like this:

```bash
 Error: ambiguity found when searching for best transformation
    Trying to produce type 'SEARCHED_LIB' from:
    Generators that succeeded:
     -  searched-lib-generator
     -  emscripten.searched-lib-generator
    First generator produced:
     -  { %.no-action-icudata.SEARCHED_LIB }
    Second generator produced:
     -  { %.no-action-icudata.SEARCHED_LIB }
```

This seems to be due to Boost searching for external libraries. This obviously happens when `boost:zlib=True` but also occurs when Boost tries to build certain libraries such as "regex", which is why those are disabled as well. [JoshS on this StackOverflow response](https://stackoverflow.com/questions/15724357/using-boost-with-emscripten) mentioned getting around this error by modifying /tools/build/src/build/generators.jam to ignore the searched-lib-generator. I bring it up in case someone more versed in the ways of bjam can figure out a way to get this to fully work.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=342) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=342) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/conan-io/conan-center-index?pullRequest=342) it.</sub>

SSE4 commented:
@TimSimpson may you submit access request here: https://github.com/conan-io/conan-center-index/issues/4 ? thanks

SSE4 commented:
@TimSimpson there are build failures, pls take a look

TimSimpson commented:
So, the build passed, but the new target I had added - which used the `timer` component instead of `exception`- experienced a linker error with Emscripten. My guess is it's due to Boost timer simply not working with Emscripten. I've changed it to use the random component instead since that obviously works.
